### ISSUE TITLE
AUR packages for ArchLinux

### ISSUE THREAD

**Issue by nazar554** [smallstep/cli#107]
Hi I made AUR packages for Arch (from GitHub release binaries). Can you mention them in `README.md`?

https://aur.archlinux.org/packages/step-ca-bin/
https://aur.archlinux.org/packages/step-cli-bin/

**Comment by mmalone** [comment]
@nazar554 this is awesome! thanks so much for your contribution!

We do have one concerned and are wondering if you have any thoughts: how can we be sure this package is kept up to date with the most recent stable builds of `step` and `step-ca`? We need a way to update it ourselves or to notify you to update when a new build is necessary. We have a couple ideas...

* We could list this as an "unofficial build" in our readme and note the version. We could open an issue or PR on your git repo when a new build is required. You could open a PR when the package is updated to change the version in the README when a new build is available, or...
* You could give us access to the arch linux repo so we can push new builds automatically. We'd probably need some assistance setting this up.

Or maybe there are other options. None of us have much experience with arch linux / pacman so suggestions are very welcome.

Thanks again!

**Comment by nazar554** [comment]
Sure I can add you as co-maintainer. You just need an AUR account. Since AUR doesn't host binary packages, it's enough to update the `PKGBUILD` with new version/checksums (using `makepkg -g`) and [`.SRCINFO`](https://wiki.archlinux.org/index.php/.SRCINFO)

**Comment by dopey** [comment]
Hey @nazar554 can you add me as a pkg maintainer? My uname on archlinux.com is `dopey`. I'll make updating the archlinux repo part of our build process. And thanks for the help in setting everything up!!

**Comment by nazar554** [comment]
@dopey Done

**Comment by dopey** [comment]
Hey @nazar554 unfortunately I'm having an issue pushing to the step-certificates repo. I looked around online but couldn't quite find what I needed.

I'm getting the following error when I try to push:
```
remote: error: The following error occurred when parsing commit
remote: error: e7f5841cf962daf90e97d278df37148b0ce67e30:
remote: error: the repository must not contain subdirectories
remote: error: hook declined to update refs/heads/master
To ssh://aur.archlinux.org/step-ca-bin.git
 ! [remote rejected] master -> master (hook declined)
error: failed to push some refs to 'ssh://aur@aur.archlinux.org/step-ca-bin.git'
```

Here are the contents of my `step-ca-bin` directory:
```
 $ la
total 32
drwxr-xr-x   7 max  staff   224 Apr 10 18:28 .
drwxr-xr-x   3 max  staff    96 Apr 10 18:28 ..
-rw-r--r--   1 max  staff   485 Apr 10 18:09 .SRCINFO
drwxr-xr-x  14 max  staff   448 Apr 10 18:28 .git
-rw-r--r--   1 max  staff   860 Apr 10 18:09 PKGBUILD
-rw-r--r--   1 max  staff   750 Apr 10 18:27 README.md
-rwxr-xr-x   1 max  staff  1662 Apr 10 18:27 update
```

The only subdirectory I see is `.git` and I think I need that one. Not sure what to make of this. I removed the new files that I added, but still no luck.

Also, I'm aware this issue isn't the right place for debugging issues with the AUR arch linux build process, but I was hoping to avoid creating a forum account (partly b/c I don't have an archlinux box at hand to run the registration question on - I know, I'm ashamed).

**Comment by nazar554** [comment]
@dopey Can you send me a a diff of what you are trying to push?

**Comment by dopey** [comment]
```
commit bb8b9e04affd784b836c77f843d2d9d43c7255f7
Author: max furman <mx.furman@gmail.com>
Date:   Wed Apr 10 17:38:47 2019 -0700

    step v0.9.2
    
    * add README
    * add update script

diff --git a/.SRCINFO b/.SRCINFO
index 60c65bb..73730b9 100644
--- a/.SRCINFO
+++ b/.SRCINFO
@@ -1,12 +1,12 @@
 pkgbase = step-ca-bin
 	pkgdesc = An online certificate authority and related tools for secure automated certificate management, so you can use TLS everywhere.
-	pkgver = 0.9.1
+	pkgver = 0.9.2
 	pkgrel = 1
 	url = https://smallstep.com/certificates
 	arch = x86_64
 	license = Apache
-	source = https://github.com/smallstep/certificates/releases/download/v0.9.1/step-certificates_0.9.1_linux_amd64.tar.gz
-	sha256sums = 67559a110d08dccfd4fc49108bb1aa58ab5c5831aae10af8dbbf4939e78f85b0
+	source = https://github.com/smallstep/certificates/releases/download/v0.9.2/step-certificates_0.9.2_linux_amd64.tar.gz
+	sha256sums = 9146f01e6ca1ffc929390d42b068108a100ddb54bf27295d0f7b49ac547e663b
 
 pkgname = step-ca-bin
 
diff --git a/PKGBUILD b/PKGBUILD
index ab694da..5117468 100644
--- a/PKGBUILD
+++ b/PKGBUILD
@@ -1,7 +1,12 @@
-# Maintainer: Nazar Mishturak <nazarmx@gmail.com> 
+# DO NOT EDIT THIS FILE DIRECTLY
+# UNLESS YOU KNOW WHAT YOU ARE DOING
+#
+# Instead use ./bin/update to update this file for new releases.
+#
+# Maintainer: dopey <max@smallstep.com>
 _binname=step-ca
 pkgname=$_binname-bin
-pkgver=0.9.1
+pkgver=0.9.2
 pkgrel=1
 pkgdesc="An online certificate authority and related tools for secure automated certificate management, so you can use TLS everywhere."
 arch=('x86_64')
@@ -9,7 +14,7 @@ url="https://smallstep.com/certificates"
 license=('Apache')
 
 source=("https://github.com/smallstep/certificates/releases/download/v${pkgver}/step-certificates_${pkgver}_linux_amd64.tar.gz")
-sha256sums=('67559a110d08dccfd4fc49108bb1aa58ab5c5831aae10af8dbbf4939e78f85b0')
+sha256sums=('9146f01e6ca1ffc929390d42b068108a100ddb54bf27295d0f7b49ac547e663b')
 
 package() {
 	install -Dm755 "step-certificates_$pkgver/bin/$_binname" "$pkgdir/usr/bin/$_binname"
diff --git a/README.md b/README.md
new file mode 100644
index 0000000..60ef5fb
--- /dev/null
+++ b/README.md
@@ -0,0 +1,27 @@
+# Smallstep Arch Linux Formula
+
+This repository contains tools for building `step certificates` in
+[Arch Linux](https://archlinux.org/).
+
+### Bumping the Version
+
+AUR will distribute the version of `step-ca` hardcoded into the
+[./PKGBUILD](./PKGBUILD). Follow the steps below to update this formula.
+
+1. Update the `step-ca-bin` package metadata.
+
+    **DO NOT** edit [./PKGBUILD](./PKGBUILD) directly. Instead run [./update](./update)
+    with the appropriate arguments to update the formula.
+
+    Select the versions of `step certificates` that will be
+    distributed in the homebrew release.
+
+    Then run:
+        ```
+        # ./update <certificates-version>
+        $ ./update v0.8.4
+        ```
+
+2. Commit and push your changes.
+
+*All Done!*
diff --git a/update b/update
new file mode 100755
index 0000000..2e4fe65
--- /dev/null
+++ b/update
@@ -0,0 +1,62 @@
+#! /bin/sh
+#/
+#/ usage: ./update <certificates-version>
+#/
+#/ positional arguments:
+#/   certificates-version   e.g. v0.8.4
+#/
+#/ example(s):
+#/   - ./update v0.8.4
+#/
+
+set -e
+
+usage() {
+    cat "$0" | grep '^#/' | cut -c 4-
+    exit $1
+}
+
+if [ $# -ne 1 ]; then
+    echo "Not enough positional arguments"
+    usage 1
+fi
+
+CERTIFICATES_URL="https://github.com/smallstep/certificates/releases/download/${1}/step-certificates_${1#?}_linux_amd64.tar.gz"
+DIR="$( cd "$(dirname "$0")" ; pwd -P )"
+
+cd "$(mktemp -d)"
+TMP="$(PWD)"
+trap "rm -rf \"${TMP}\"" EXIT INT QUIT TERM
+
+curl -o "certificates.tar.gz" -J -L -s "${CERTIFICATES_URL}"
+CERTIFICATES_SHA="$(shasum -a 256 "certificates.tar.gz" | awk '{printf $1}')"
+
+VERSION=$(echo ${1} | sed 's/^v//')
+
+cat > "${DIR}/PKGBUILD" <<EOF
+# DO NOT EDIT THIS FILE DIRECTLY
+# UNLESS YOU KNOW WHAT YOU ARE DOING
+#
+# Instead use ./bin/update to update this file for new releases.
+#
+# Maintainer: dopey <max@smallstep.com>
+_binname=step-ca
+pkgname=\$_binname-bin
+pkgver=${VERSION}
+pkgrel=1
+pkgdesc="An online certificate authority and related tools for secure automated certificate management, so you can use TLS everywhere."
+arch=('x86_64')
+url="https://smallstep.com/certificates"
+license=('Apache')
+
+source=("https://github.com/smallstep/certificates/releases/download/v\${pkgver}/step-certificates_\${pkgver}_linux_amd64.tar.gz")
+sha256sums=('${CERTIFICATES_SHA}')
+
+package() {
+	install -Dm755 "step-certificates_\$pkgver/bin/\$_binname" "\$pkgdir/usr/bin/\$_binname"
+	install -Dm644 "step-certificates_\$pkgver/README.md" "\$pkgdir/usr/share/doc/\$pkgname/README.md"
+}
+EOF
+
+cd ${DIR}
+makepkg --printsrcinfo > "${DIR}/.SRCINFO"
```

* updated versions and shasums in `PKGBUILD` and `.SRCINFO`
* added `README`
* added `update` script for for autoupdating metadata files in the future.

**Comment by nazar554** [comment]
@dopey 
I don't think you should keep the `update` script in AUR repo, as it is not part of `sources` array.
You also don't need to update package sums manually, it's enough to change `pkgver` in `PKGBUILD` and run `updpkgsums` from `pacman-contrib`.  It will download the binaries and update checksums automatically.

Maybe AUR repo can be added as git submodule and `update` script can be in top-level repo.

It can look like
```bash
sed -i "s/^pkgver=.*/pkgver=${VERSION}/" PKGBUILD
updpkgsums
makepkg --printsrcinfo > .SRCINFO
git commit -m "Updated to $VERSION" && git push

```

**Comment by dopey** [comment]
hey @nazar554. 

It's weird because I actually removed the README and `update` script as a test and still wasn't able to push before. 🤷‍♂️ 

Thanks for the tip on using submodules - I ended up implementing the flow almost exactly this way, so I appreciated the code snippet as well. I'll usually be running on mac so I replaced `updpkgsums` with the hacky manual way of doing it that you saw before. 

I've updated both repos with the newest release (and updated the `install` section of the README here) however now I'm having trouble testing the packages. Never used arch linux before so this will be a learning process :) I'll hit the newbie boards.

Thanks for your help and guidance! Lmk if the installation instructions are insufficient.
